src/main/java/org/springframework/integration/kafka/outbound/KafkaProducerMessageHandler.java

```diff
@@ -142,28 +144,15 @@ protected void handleMessageInternal(final Message<?> message) throws Exception
 				? this.messageKeyExpression.getValue(this.evaluationContext, message)
 				: message.getHeaders().get(KafkaHeaders.MESSAGE_KEY);
 
-		ListenableFuture<?> future;
+		Long timestamp = message.getHeaders().get(KafkaHeaders.TIMESTAMP, Long.class);
```

Comment:
@artembilan, I have updated the PR now. I have very likely missed some of your points, if you can have a quick review and let me know how it looks now.